### ISSUE TITLE
ability to set timeout within step function body

### ISSUE THREAD

**Issue by tommysullivan** [cucumber/cucumber-js#537]
sometimes we wish to have a dynamically determined timeout, such as one expressed within the Gherkin itself and matched on; or sometimes based on some other dynamic behavior that occurs within the step definition function body itself.

I propose something like callback.timeout(10000), which would have to be called before any default timeout (global or as 2nd param to the step definition {timeout:x} form). If it is set before that, then the default timeout is cancelled and a new timeout set where the timer starts at that instant.

The other thing is that in IntelliJ the plugin does not work when the { timeout: x } 2nd param is used, and one cannot jump from Gherkin to the step definition. Solving for dynamic timeout would indirectly create a workaround for the plugin deficiency.

**Comment by charlierudolph** [comment]
I'm not a big fan of this. Can you give more detail about your use case? If you are just passing in a wait time, you can just set the timeout to be higher then any value passed in by your features.

**Comment by tommysullivan** [comment]
The use case is that I have SLAs on various scenarios that are part of the product requirements, so I want to capture them in Gherkin. Then, in my step definition, I would like to consume the timeout duration from the gherkin, and set up the behavior that if the test does not resolve to a pass or fail before the timeout duration has passed, then the test is marked as a failure for timing out.

Since by definition I do not know what the timeout will be in the Gherkin, I cannot set the step definition timeout to be higher than the largest timeout in the Gherkin.

There are ways to work around this but I was wondering if it might be useful to have framework support for this kind of use case.

**Comment by tommysullivan** [comment]
Just to be clear, I am working on the premise that the Gherkin is controlled mostly by the product team and they can arbitrarily change the SLAs; I simply write the step definitions to consume whatever given SLA they have written in their Gherkin.

It is easy to just assume a really long timeout and then use the shorter timeout in the steps themselves. It just feels a little clunky. And sometimes I make an error and then the test takes a really long time to timeout and I wish the framework had helped prevent that.

**Comment by charlierudolph** [comment]
You could set the timeout for the actual step as higher then any value that would be passed in and then use the value passed in from the step to setup your own timeout.
```js
this.Given(/x is performed in (\d) seconds/, {timeout: 30000}, function (maxWaitTime, callback) {
   if (maxWaitTime > 30) {
     throw new Error('To wait for more than 30 seconds, increase the step timeout');
   };
   setTimeout(funciton() { 
     callback('x was not performed in ' + maxWaitTime + ' seconds');
   }, maxWaitTime * 1000);
   // poll to see if is x is perfomed 
});
```
The step timeout is meant more for a guard to ensure the step eventually completes and doesn't hang forever. When I've used selenium webdriver, it has its own timeout for how long to wait for a specific element. I think its better to have your own timeout so you can give better error messages for what went wrong instead of just getting the message "Step timed out". What do you think?

**Comment by tommysullivan** [comment]
Yeah totally, its just that I am returning promises, and so I have to like create an "all" promise over the operation's promise and a separate "timeout promise". And then when anything goes wrong, the step runs for the huge timeout duration before finally giving an error. You know what I mean?

**Comment by tommysullivan** [comment]
So it would be so concise and straightforward to just be like this.timeout(2000); and call it a day, just return my normal promise, and not worry about it ever again. And then as a side benefit intellij plugin would not get all confused by the presence of the {options} and would allow me to navigate from the features to the step definitions in a single click. But that is a tangent.

**Comment by charlierudolph** [comment]
If you are implementing your own timeout based on a value passed in, I don't see why you have to wait a huge timeout duration. Can you please provide some sample code? Maybe give your current code and what you would like the code to look like in the ideal case.

**Comment by tommysullivan** [comment]
Sure. Current Code example:

`
this.Given('something', {timeout: 10000000}, function(slaTimeout) {
   var timeoutPromise = new Promise((resolve, reject)=>{
      setTimeout(reject, slaTimeout);
   });
   var opUnderTestPromise = this.someAsyncOperationIWantToTest();
   var racePromise = Promise.race([timeoutPromise, opUnderTestPromise]);
   return expect(racePromise).to.eventually.equal(7);
});
`

Here's what I'd like:
`
this.Given('something', function(slaTimeout) {
    this.setTimeout(slaTimeout);
    return expect(this.someAsyncOperationIWantToTest()).to.eventually.equal(7);
});
`

**Comment by tommysullivan** [comment]
you can see that in the current code example, it is easier, especially for a novice, to mess things up with the slaTimeout, and then the framework will of course wait for the really long timeout before failing.

**Comment by charlierudolph** [comment]
If you create a function to deal with timing out a promise it can almost as simple.
Bluebird promises actually come with a [timeout](http://bluebirdjs.com/docs/api/timeout.html) function too.
```js
function addTimeoutToPromise(promise, timeout, reason) {
   var rejectOnTimeout = new Promise((resolve, reject)=>{
      setTimeout(function() { 
        reject(reason) 
      }, timeout);
   });
   return Promise.race([rejectOnTimeout, promise]);
}

this.Given('something', function(slaTimeout) {
  var promise = addTimeoutToPromise(this.someAsyncOperationIWantToTest(), slaTimeout, 'some async operation timed out');
  return expect(promise).to.eventually.equal(7);
});
```
You could probably extend whatever promise object you are working with to have the timeout functionality similar to bluebird. 

I really think there is value in using more fine grained control and timing out specific actions in order to give yourself more specific error messages.

**Comment by tommysullivan** [comment]
Yeah sure, people can all just do that whenever they want this behavior. I could easily see a parallel universe in which the API for setting timeouts had this capability and it would be totally handy and useful, as an additional facility beyond the finer tuned control u refer to, which often is actual musky a good idea :)

**Comment by tommysullivan** [comment]
Lol usually a good idea*

**Comment by charlierudolph** [comment]
What do you think about just allowing timeout to be a function that gets passed in the step parameters?
Keeps the same interface in the options object but allows it to be more dynamic based on the parameters.
```js
getTimeout = function (action, slaTimeout) { 
  return slaTimeout * 1000 
};
this.Given(/(.*) is performed in (\d) seconds/, {timeout: getTimeout}, function(action, slaTimeout) {
   return expect(this.someAsyncOperationIWantToTest()).to.eventually.equal(7);
});
```

**Comment by tommysullivan** [comment]
That's a cool idea, it wouldn't pollute the "this". Yeah maybe!

**Comment by charlierudolph** [comment]
Would you like to take a crack at implementing this?

**Comment by tommysullivan** [comment]
I can try and take a stab in the next couple weeks for sure. Big backlog at the moment. I built a nice angular result viewer that I was thinking of open sourcing. Thanks for the feedback

**Comment by JimTheMan** [comment]
In general, putting timeouts in unit tests is a bad idea. Instead, you should mock thimeout.

**Comment by charlierudolph** [comment]
@tommysullivan closing this due to age. An implementation of what we discussed would still be accepted.

**Comment by charlierudolph** [comment]
Thinking a bit more on this, I no longer like what we discussed. The step timeout error is non-informative and doesn't let you know the reason for the step failure. The step timeout error was built to catch cases where the callback interface was used and forgot to execute the callback or for when a step is clearly taking too long. 

I think my earlier [suggestion](#issuecomment-215975713) is the way to go.

**Comment by tommysullivan** [comment]
The legitimate business case, regardless of implementation details, is: 

1. Business Owner can update Gherkin to document timeouts as part of SLA specification
2. Business Owner can continue to set the global timeout in the world file
3. A mechanism enables the Gherkin-derived timeout to override the global step timeout

Let us first see whether we acknowledge the legitimacy of this as a use case of Gherkin / Cucumber.

**Comment by charlierudolph** [comment]
Part 3 is the only spot i have a problem. I disagree that Gherkin derived timeout should override the global step timeout. 

The step timeout is a fail safe and it erring gives you no information about the actual cause. Something internal may have taken too long or you may have forgotten to execute the passed in callback. 

Setting your own timeout for specific pieces of the code gives you fine grained control and better error messages. Thus the step should have its timeout configured to be large enough to cover any case or step timeouts should be disabled (that would be a new feature)

**Comment by tommysullivan** [comment]
> The step timeout is a fail safe and it erring gives you no information about the actual cause. 

Totally understood.

> Something internal may have taken too long or you may have forgotten to execute the passed in callback.

Also understood.

> Setting your own timeout for specific pieces of the code gives you fine grained control and better error messages.

Absolutely. I do not want to use the feature I am requesting to actually express the testing time limits. I will use actual expectations for that for the reasons you mentioned. 

> Thus the step should have its timeout configured to be large enough to cover any case or step timeouts should be disabled (that would be a new feature)

Yes exactly! The step needs a timeout that will suffice for the given step in question. Since the time that the step in question needs is derived from the text of the Gherkin step, any solution which does not allow me to override the timeout, just for that step, based on the gherkin content, will not work.

Since this capability does not exist, I am forced to leverage one of the following alternatives, which all have problems:

Workaround 1
Increase the global timeout to be longer than the longest test.

This is a problem because (a) it requires that if my product person updates the SLAs in the gherkin, that they (or someone else) needs to go and update the configured global timeout to be longer than that; and (b) with a large global timeout, I lose (for all practical purposes) the "fail safe" that it gives me for those tests where  "something internal may have taken too long or you may have forgotten to execute the passed in callback."

Workaround 2
Set the timeout using the step definition extra arguments

This is a problem because (a) it requires that if my product person updates the SLAs in the gherkin, that they (or someone else) needs to go and update the timeout in the corresponding steps. This can be changed if I force the SLAs into a configuration file somewhere else, and then have the Gherkin refer to that configuration as well as have the steps refer to it. But that kind of undermines the philosophy of having the Gherkin be a self-contained documentation system. I want people to review the test results and read them and understand exactly what happened with minimal indirection.

If this is a legitimate issue, I don't know that it is a good idea to close it. I would love to find the time to help out. This recently came up again internally as a need from another of our QA, so perhaps that will help me justify using company time to try and code up a solution. My concern is that I might have to take quite a bit of time to try and understand your code in order to add this enhancement to it. I might be interested in sending a little cash your way if you could implement a solution or know someone who is familiar with the codebase who could do it.

**Comment by tommysullivan** [comment]
Scenario: Test an SLA
Given ...
When I ...
Then the operation completes within "25" minutes

Scenario: Test an SLA Part Deux
Given ...
When I ...
Then the operation completes within "2" minutes

---

@Then(^...)
verifyOperationCompletionTime(minutes:string):PromisedAssertion {
  //it is only HERE when I have the minutes that I can know what the step timeout should be.
  //note that I would not use the step timeout to express the expectation though!! I would only
  //override it so that my test case does not get interrupted early by your framework.
  return expect(promiseFromEarlierOperation).to.eventually.be.fulfilled.in(5).minutes;
}

**Comment by tommysullivan** [comment]
With that example in mind, I might want something like:

```
@Then(^...)
verifyOperationCompletionTime(minutes:string):PromisedAssertion {
  this.setStepTimeout(parseInt(minutes)*60000 + 5000);
  return expect(promiseFromEarlierOperation).to.eventually.be.fulfilled.in(minutes).minutes;
}

**Comment by tommysullivan** [comment]
Then my error message would be nice:

```
Then the operation completes within "2" minutes - failed - wanted 2 minutes but got 2 minutes and 3 seconds.
```

**Comment by tommysullivan** [comment]
Hopefully you can see with this example that setting the timeout for the single step definition using the existing API would require me setting it to at least 25 minutes; if someone updated the SLA for a longer operation then it would fail. I'd have to make a code change. Meantime if there is an unintentional error somewhere in one of the shorter operations, then it would take an absurdly long time before the framework would catch it, causing the test run as a whole to be way too long.

This is a legitimate business use case.

**Comment by charlierudolph** [comment]
For this case, I'd rather just implement the ability to disable the step timeout (by passing in `{timeout: -1}`) and require you to implement your own timeout for that specific step.

**Comment by tommysullivan** [comment]
Oh totally, that would work wonderfully!!!

**Comment by charlierudolph** [comment]
Released in 2.0.0-rc.6
